Wide HTML5 `<video` and `<iframe` Support

hnryjms opened the pull request:
Hi there 👋

I wanted to do a workout on Les Mills On Demand with some friends, and found this plugin. I originally aimed to just build a new connector for that site when I forked the project ... but I ended up running into a bunch of issues and didn't know how to use the old approach for their DOM structure.

I also noticed that a lot of the code between the YouTube and Prime Video connectors were similar, so I restarted the project from the basis that any mostly-standard HTML5 video element will be supported. This way clicking on the Extension plugin generates the URL as long as you're on any compatible video site.

Here are my changes:

- WatchParty: Support for most HTML5 `<video` players
  - [x] Upgraded to ES6 modules via https://github.com/otiai10/chrome-extension-es6-import
  - [x] Use `hash` instead of `queryParams` in URL (hide watchparty from server)
  - [x] Replaced JQuery w/ native event handler logic
  - [x] Improved Chrome.tabs interaction w/ primary tab argument
  - [x] Clear multi-session support with session merging
- WatchParty: IFrame support with nested monitoring
  - [x] MutationObserver code for watching & waiting for `<video`
  - [x] Consolidated loader scripts into separate directory
  - [x] In-Frame code relies on background task for runtime events
  - [x] Session connect/disconnect tracking & browser notification
  - [x] Monitors all `<iframe` content for `<video` tag continuously
- WatchParty: Fix for YouTube hash-change refresh problem
  - [x] Use `history.replaceState()` for editing hash marker, which avoids a browser refresh (fix for YouTube)
  - [x] Fixed hashComponents legacy usage from pre-abstraction
  - [x] Improved tab-used-to-have-video & non-active-tab-has-video error messages
  - [x] Added notification for Host leaving the session
  - [x] Improved code documentation for event transmit loop avoidance
- WatchParty: Fixed PeerJS Reconnect `id` Errors
  - [x] All events are shared via WebRTC, including caused by internal actions. This adds support for a host propagating it's events onward to adjacent watching peers.
  - [x] Simplified MutationObserver to rescan for `video` tag after mutations are complete.
  - [x] Improved error handling for peer & session. Hosts ignore session errors since it can still serve other peers. Added monitoring for both outgoing-connection (peer) and incoming-connection (session) errors.

hnryjms commented:
Not really sure what to do about this tbh.

If you want to dissolve your repo into mine, I can be the maintainer of WatchParty (and esp. the Chrome Web Store)....or just give you credit and have you manage the Web Store entry 🤷‍♀️

hnryjms commented:
Also... I tested this in YouTube at 947e1de03cd4e2b1a991dd731cb0787f796f5523 but not since 13aa8fb92bc0f77dc2f4de0febc3ebc306197a65

I'll do more extensive testing the next time I'm free. Hoping to use this with my friends on Saturday.

mrsheepsheep commented:
Wow ! I didn't expect someone to find my repo !
This was originally a prototype, I didn't even plan to release it on the webstore, but we can discuss about this, by email, discord, other ? I'm interested 
You did some crazy work here, thanks a lot.
Testing this asap, I'll keep you updated !
(I probably messed up your checkboxes sorry about that)

hnryjms commented:
Discord: `hnryjms#2305` — Send me a DM!

hnryjms commented on action.js:
```diff
@@ -1,10 +1,96 @@
-        });
+// We monitor all `<iframe` and top-frame content for a `<video` tag, and store
+// the frame info to be used when our chrome.browserAction is clicked.
+let videoFrame;
```
I have a feeling that `videoFrame` is going to retain the old frame even after the user closes the tab/navigates away. That probably causes the extension to crash if you click the Extension button on a page with no video.

Maybe should address that (?)

hnryjms replied:
This was true. If your tab ever shows a video and then stops, it would host a session and get into a weird state.

It's pretty hard to know when to clear this value, so I'm just gonna throw a vague error if you press the button

hnryjms commented:
New discovery: `window.location.href = 'xyz'` causes YouTube to refresh, which disconnects the session. Using `history.replaceState(null, null, '#xyz');` fixes that.

hnryjms commented on src/js/session.js:
```diff
@@ -0,0 +1,92 @@
+        chrome.runtime.sendMessage({
+            type: 'sessionEnded',
+            error
+        })
```
🤷‍♀️ aint no js developers got time to learn native OS memory handling.

hnryjms commented:
Okay I think it's ready to try out now. Got it working in YouTube and Les Mills On Demand. Gonna test it with some more websites ... maybe (?)

mrsheepsheep commented:
I hope to be able to go more in depth into your changes next week. I added you on Discord !
I tried youtube at https://github.com/mrsheepsheep/watchparty/commit/13aa8fb92bc0f77dc2f4de0febc3ebc306197a65 and indeed encountered https://github.com/mrsheepsheep/watchparty/pull/1#issuecomment-618763551 
I'll ask some friends to try it out too so they can give some feedback too

SudharsanSukumar commented:
Hey I'm also interested in contributing! Discord is `MustyRusty#6090`.

SudharsanSukumar commented:
Found a bug in dom-wait.js when testing Netflix. It only searches top level nodes. In the case of netflix, the video element is inserted inside of a div. Suggested fix via dfs that works (minimal testing). Sorry if the format is messed up for some reason it's not working and I don't feel like messing with it longer than I already have  🤣 

`// See https://gist.github.com/PaulKinlan/2d7cd4e78a63a97387137a0a9fb7ee6e
export function waitForElement(selector) {
  return new Promise(function (resolve) {
    const element = document.querySelector(selector);

    if (element) {
      resolve(element);
      return;
    }

    const observer = new MutationObserver(function (mutations) {
      mutations.forEach(function (mutation) {
        const nodes = Array.from(mutation.addedNodes);
        const matchedNode = searchNodesForMatch(nodes, selector);
        if (matchedNode) {
          observer.disconnect();
          resolve(matchedNode);
        }
      });
    });

    observer.observe(document.documentElement, {
      childList: true,
      subtree: true,
    });
  });
}

function searchNodesForMatch(nodes, selector) {
  for (const node of nodes) {
    if (node.matches && node.matches(selector)) {
      return node;
    }

    if (node.childNodes) {
      return searchNodesForMatch(node.childNodes, selector);
    }
  }

  return null;
}
`

hnryjms commented:
I also discovered that after about 10 minutes, the watcher disconnects from the `<video` tag on Les Mills On Demand. I'm thinking that somehow, the website unmounts the original `<video` DOMElement and mounts a new one without interrupting the play.

I'm thinking of improving the `MutationObserver` aspect of the code so that it not only finds the video on first load, but keeps track of the video as the page changes rendering ... because it doesn't have the listeners on second/third mount of the video, and so you end up "losing your connection" even though PeerJS stays open

hnryjms commented:
Hey team, I just pushed some new changes.

1. Updated PeerJS seems smart enough to not send our own messages. I removed the session-sender-monitoring checks that are just adding complexity now.
1. I found that a 3+ session would only synchronize if events are performed by the host. Removing the `#lastEvent` tracking means that events are **always** emitted outward...even if it's mostly a re-transmission. But because we allow `0.25` seconds of `currentTime` leniency, users with fast connections shouldn't notice anything. I *think* this also means if one person's video pauses to buffer, it'll pause everyone's video (which is probably desired behavior?)
1. Some better error handling for problems occurring on both `host/peer` and `session` ... which represent outgoing-connection and incoming-connection respectively.
1. I also took @SudharsanSukumar's suggestion on re-scanning the DOM after a mutation ... but made it a bit simpler by just assuming that redoing `document.querySelector` will just give us what we want. The sites I'm testing with always have the `<video` element in the DOM at load (aka they're not client-side-JS), so I don't know how to test this.

Try it out and let me know if it works better for you!

SudharsanSukumar commented:
Testing the latest

1. YouTube doesn't seem to be working at all for me. Clicking the extension either doesn't do anything or opens the alert when a video isn't detected.
2. Netflix seems to be working (at least for hosting haven't tested multiple users.

Any idea why YouTube is having issues now?

mrsheepsheep commented:
Hi guys
I'm currently working in the vines (literally) to help during the covid, and being out there all day long I'm not sure I'll be able to follow your rythm !
@hnryjms I sent you an invite so I can give you the Maintainer role
@SudharsanSukumar thanks for joining !
The gecko stuff in the manifest was initially there to load the extension inside Firefox, but PeerJS was not working properly last time I tried.
Consider removing it for now.
